File: external/cnem_03-10-17/python/tuto/tuto_interpol2d.py
##
## Copyright (C) 2003-2011
## Lounes ILLOUL (illoul_lounes@yahoo.fr)
## Philippe LORONG (philippe.lorong@ensam.eu)
## Arts et Metiers ParisTech, Paris, France
##
## $Revision$ - $Date$

import shutil
import re
import math
import cPickle
import numpy as np
import scipy as sp

import sys
sys.path.append('../bin')
import load_node_front
import py_cnem2d

#-------------------------------------------------------------------------------
# read nodes and boundary :
print '\n--------------------------------------------------------------------------------\n'
print 'read node and boundary...\n'

[XY_Noeud,Nb_Noeud_Front,Ind_Noeud_Front]=load_node_front.load2d('data2d_big/')

#-------------------------------------------------------------------------------
# creat dummy point location for interpolation evaluation

XY_Point=list((np.random.random(100000*2)-0.5)*1.5+0.5)

#-------------------------------------------------------------------------------
# creat the interpolator
print '\n--------------------------------------------------------------------------------\n'
print 'creat the interpolator...\n'

interpol=py_cnem2d.interpol(XY_Noeud,Nb_Noeud_Front,Ind_Noeud_Front,XY_Point)

#----------------------------------------------------------------------------
# interpolate variable filds : Var
print '\n--------------------------------------------------------------------------------\n'
print 'interpolate variable filds...\n'

# test : Var = XY_Noeud ==> interpolated Var on XY_Point = XY_Point

Var=np.array(XY_Noeud)
Var=Var.reshape((Var.shape[0]/2,2))

#interpolate Var :
Var_Int=interpol.interpolate(Var)

XY_Point=np.array(XY_Point)
XY_Point=XY_Point.reshape((XY_Point.shape[0]/2,2))

Dif=Var_Int-XY_Point

maxdif=0.

for i in xrange(len(Dif)):
    if interpol.In_Out[i] : # point interpolation in the domain 
        maxdif_i=max(abs(Dif[i]))
        if maxdif_i>maxdif:
            maxdif=maxdif_i
        
print 'err max : ', maxdif

#-------------------------------------------------------------------------------

File: external/cnem_03-10-17/matlab/tuto2d/tuto_interpol2d.m
% Copyright (C) 2003-2011
% Lounes ILLOUL (illoul_lounes@yahoo.fr)
% Philippe LORONG (philippe.lorong@ensam.eu)
% Arts et Metiers ParisTech, Paris, France

clc;
clear all;
close all;
addpath('../bin');

%% load domain (nodes + boundary(s))

load('data_small'); 

%% creat dummy point location for interpolation evaluation

XY_Point=(rand(1000,2)-0.5)*1.5+0.5;

%% plot 

% boundarys

figure;
hold on;
axis equal;

plot_boundary(XY_Node,Nb_Node_Front,Ind_Node_Front);

% node

plot(XY_Node(:,1),XY_Node(:,2),'.','color','green');

% point

plot(XY_Point(:,1),XY_Point(:,2),'o','color','blue');

hold off;

%% initialize the interpolator

Interpol=m_cnem2d_interpol(XY_Node,Nb_Node_Front,Ind_Node_Front,XY_Point);

%% interpolate a filds
% test : Var = XY_Noeud ==> interpolated Var on XY_Point = XY_Point

Var=XY_Node;
Var_Int=Interpol.interpolate(Var);

%% cal error

dif=Var_Int-XY_Point;
j=0;
ind_p_in=zeros(sum(Interpol.In_Out),1);
for i=1:size(Interpol.In_Out,1)
    if Interpol.In_Out(i)
        j=j+1;
        ind_p_in(j)=i;
    end
end
int_p_out=setdiff(1:size(Interpol.In_Out,1),ind_p_in);

err=max(max(abs(dif(ind_p_in,:))))

%% plot in/out point

% boundarys

figure;
hold on;
axis equal;

j=1;
for i=1:size(Nb_Node_Front,1)
    XY_Node_Front_i=XY_Node(Ind_Node_Front([j:(j+Nb_Node_Front(i)-1)]),:);
    plot(XY_Node_Front_i(:,1),XY_Node_Front_i(:,2),'color','black','LineWidth',2);
    j=j+Nb_Node_Front(i);
end

% in

plot(XY_Point(ind_p_in,1),XY_Point(ind_p_in,2),'*','color','red');

% out

plot(XY_Point(int_p_out,1),XY_Point(int_p_out,2),'+','color','blue');

hold off;

%%